File: tools/forge_lab/policies.py
"""
Imperfect player policies for Monte Carlo difficulty estimation.

Each policy plays a level to win/fail. Running many policies many times gives the
empirical win-rate that defines how hard a level actually FEELS - distinct from the
solver, which only proves a winning line exists.

Actions are unified: ("lane", i) dispatches a queue front; ("buffer", dish)
re-dispatches a dock plate (a move). Policies can park plates in the dock and send
them back later, matching the shipping engine.
"""

from __future__ import annotations

import random

from forge_sim import (
    Level,
    State,
    can_advance,
    dispatch_actions,
    apply_action,
    initial_state,
    is_victory,
    tick,
)


def _wanted_dishes(s: State) -> set[str]:
    return {seat.dish for seat in s.seats if seat.customer_idx >= 0 and seat.turnover == 0 and seat.dish}


def _dish_of(s: State, action: tuple) -> str:
    kind, val = action
    return s.queues[val][0] if kind == "lane" else val


def _choose(s: State, level: Level, policy: str, rng: random.Random):
    """Return a chosen action tuple, or None (meaning: tick / wait)."""
    acts = dispatch_actions(s)
    if not acts:
        return None
    wanted = _wanted_dishes(s)

    if policy == "random":
        if rng.random() < 0.7:
            return rng.choice(acts)
        return None

    if policy == "greedy":
        # Prefer delivering a wanted dock plate (frees the dock), then a wanted lane.
        dock = [a for a in acts if a[0] == "buffer" and a[1] in wanted]
        if dock:
            return rng.choice(dock)
        good = [a for a in acts if a[0] == "lane" and _dish_of(s, a) in wanted]
        if good:
            return rng.choice(good)
        return None

    if policy in ("lookahead1", "lookahead2"):
        good = [a for a in acts if (_dish_of(s, a) in wanted)]
        if good:
            def score(a: tuple) -> int:
                nxt = tick(apply_action(s, level, a), level)
                return (1000 if nxt.failed else 0) + len(nxt.buffer)
            return min(good, key=score)
        return None

    return None


def play(level: Level, policy: str, rng: random.Random, max_ticks: int = 4000) -> bool:
    """Play one game with the given policy. Returns True on victory."""
    s = initial_state(level)
    ticks = 0
    while not is_victory(level=level, s=s) and not s.failed and ticks < max_ticks:
        action = _choose(s, level, policy, rng)
        # When nothing useful and belt is idle, park a plate to make progress.
        if action is None and not can_advance(s):
            acts = dispatch_actions(s)
            if acts:
                action = acts[0]
        if action is not None:
            s = apply_action(s, level, action)
            continue
        if can_advance(s):
            s = tick(s, level)
            ticks += 1
        else:
            break
    return is_victory(s=s, level=level)


def play_trace(level: Level, policy: str, rng: random.Random, max_ticks: int = 4000):
    """Like play(), but returns (won, moves) where moves is the dispatch order."""
    s = initial_state(level)
    ticks = 0
    moves: list[dict] = []
    while not is_victory(level=level, s=s) and not s.failed and ticks < max_ticks:
        action = _choose(s, level, policy, rng)
        if action is None and not can_advance(s):
            acts = dispatch_actions(s)
            if acts:
                action = acts[0]
        if action is not None:
            moves.append({
                "index": len(moves) + 1,
                "lane": action[1] if action[0] == "lane" else -1,
                "source": "queue" if action[0] == "lane" else "buffer",
                "dish": _dish_of(s, action),
                "tick": s.tick,
            })
            s = apply_action(s, level, action)
            continue
        if can_advance(s):
            s = tick(s, level)
            ticks += 1
        else:
            break
    return is_victory(s=s, level=level), moves


def win_rate(level: Level, policy: str, runs: int, rng: random.Random) -> float:
    wins = sum(1 for _ in range(runs) if play(level, policy, rng))
    return wins / runs

File: src/forge/policies.ts
/**
 * Player policies + Monte Carlo playouts (in-browser).
 *
 * These let the generator:
 *   - estimate a level's *felt* difficulty by simulating many naive players,
 *   - validate solvability when the exact solver caps out (large/hard levels),
 *   - extract a concrete "follow this order" move sequence from a winning policy.
 *
 * A "simulation run" = one full playout from the initial state to win/lose.
 */

import {
  GameState,
  LevelDef,
  Action as DispatchAction,
  createInitialState,
  applyAction,
  tick,
  isVictory,
  dispatchActions,
} from './simulator';
import { DishLetter, MovePlanStep } from './types';

export type PolicyName = 'random' | 'greedy' | 'lookahead';

const STEP_GUARD = 6000;

function canAdvance(s: GameState): boolean {
  return (
    s.belt.length > 0 ||
    s.seats.some((seat) => seat.turnover > 0) ||
    s.queues.some((lane) => lane.length > 0)
  );
}

/** Dishes currently wanted by a seated, eating customer. */
function wantedSet(s: GameState): Set<DishLetter> {
  const set = new Set<DishLetter>();
  for (const seat of s.seats) {
    if (seat.customerIdx >= 0 && seat.turnover === 0 && seat.dish) set.add(seat.dish);
  }
  return set;
}

type Action = { type: 'act'; action: DispatchAction } | { type: 'tick' } | null;

function dishOf(s: GameState, a: DispatchAction): DishLetter {
  return a.kind === 'lane' ? (s.queues[a.lane][0] as DishLetter) : a.dish;
}

function greedyAction(s: GameState): Action {
  const acts = dispatchActions(s);
  const wanted = wantedSet(s);
  // Prefer delivering a wanted dock plate (frees the dock), then a wanted lane.
  const wantedDock = acts.find((a) => a.kind === 'buffer' && wanted.has(a.dish));
  if (wantedDock) return { type: 'act', action: wantedDock };
  const goodLane = acts.find((a) => a.kind === 'lane' && wanted.has(dishOf(s, a)));
  if (goodLane) return { type: 'act', action: goodLane };
  if (s.belt.length > 0 || s.seats.some((seat) => seat.turnover > 0)) return { type: 'tick' };
  const laneActs = acts.filter((a) => a.kind === 'lane');
  if (laneActs.length > 0) return { type: 'act', action: laneActs[0] }; // park to progress
  if (canAdvance(s)) return { type: 'tick' };
  return null;
}

function randomAction(s: GameState, rng: () => number): Action {
  const acts = dispatchActions(s);
  const advance = canAdvance(s);
  if (acts.length === 0) return advance ? { type: 'tick' } : null;
  if (advance && rng() < 0.45) return { type: 'tick' };
  return { type: 'act', action: acts[Math.floor(rng() * acts.length)] };
}

/** Score a state for lookahead: progress made (fewer plates left) minus buffer pressure. */
function scoreState(s: GameState): number {
  const platesLeft =
    s.queues.reduce((a, l) => a + l.length, 0) +
    s.belt.length +
    s.seats.reduce((a, seat) => a + Math.max(0, seat.remaining), 0);
  return -platesLeft - s.buffer.length * 3 - (s.failed ? 1000 : 0);
}

/** Roll forward greedily for a short horizon and return the resulting score. */
function greedyRollout(start: GameState, level: LevelDef, horizon: number): number {
  let s = start;
  for (let i = 0; i < horizon; i++) {
    if (s.failed || isVictory(s, level)) break;
    const a = greedyAction(s);
    if (!a) break;
    s = a.type === 'tick' ? tick(s, level) : applyAction(s, level, a.action);
  }
  return scoreState(s) + (isVictory(s, level) ? 500 : 0);
}

function lookaheadAction(s: GameState, level: LevelDef): Action {
  const candidates: Action[] = dispatchActions(s).map((action) => ({ type: 'act' as const, action }));
  if (canAdvance(s)) candidates.push({ type: 'tick' });
  if (candidates.length === 0) return null;

  let best: Action = candidates[0];
  let bestScore = -Infinity;
  for (const a of candidates) {
    if (!a) continue;
    const next = a.type === 'tick' ? tick(s, level) : applyAction(s, level, a.action);
    if (next.failed) continue;
    const score = greedyRollout(next, level, 14);
    if (score > bestScore) {
      bestScore = score;
      best = a;
    }
  }
  return best;
}

function actionFor(s: GameState, level: LevelDef, policy: PolicyName, rng: () => number): Action {
  switch (policy) {
    case 'random':
      return randomAction(s, rng);
    case 'greedy':
      return greedyAction(s);
    case 'lookahead':
      return lookaheadAction(s, level);
  }
}

export interface PlayoutResult {
  won: boolean;
  bufferPeak: number;
  moves: MovePlanStep[];
}

/** Play one full game with the given policy. Records the dispatch sequence. */
export function playout(level: LevelDef, policy: PolicyName, rng: () => number): PlayoutResult {
  let s = createInitialState(level);
  let bufferPeak = s.buffer.length;
  const moves: MovePlanStep[] = [];
  let guard = 0;

  while (!isVictory(s, level) && !s.failed && guard++ < STEP_GUARD) {
    bufferPeak = Math.max(bufferPeak, s.buffer.length);
    const a = actionFor(s, level, policy, rng);
    if (!a) break;
    if (a.type === 'tick') {
      s = tick(s, level);
    } else {
      const available = dispatchActions(s).length;
      moves.push({
        index: moves.length + 1,
        lane: a.action.kind === 'lane' ? a.action.lane : -1,
        source: a.action.kind === 'lane' ? 'queue' : 'buffer',
        dish: dishOf(s, a.action),
        tick: s.tick,
        availableChoices: available,
        validChoices: 1,
        errorChance: 0,
        critical: false,
      });
      s = applyAction(s, level, a.action);
    }
  }
  bufferPeak = Math.max(bufferPeak, s.buffer.length);
  return { won: isVictory(s, level), bufferPeak, moves };
}

/** Empirical win-rate over `runs` playouts of a policy. */
export function winRate(
  level: LevelDef,
  policy: PolicyName,
  runs: number,
  rng: () => number,
): { rate: number; runs: number } {
  let wins = 0;
  for (let i = 0; i < runs; i++) {
    if (playout(level, policy, rng).won) wins++;
  }
  return { rate: runs ? wins / runs : 0, runs };
}

/**
 * Validate solvability via strong policies (used when exact search caps).
 * Returns a winning move sequence if found.
 */
export function policySolve(
  level: LevelDef,
  rng: () => number,
): { solvable: boolean; moves: MovePlanStep[]; bufferPeak: number } {
  for (const policy of ['lookahead', 'greedy'] as PolicyName[]) {
    const res = playout(level, policy, rng);
    if (res.won) return { solvable: true, moves: res.moves, bufferPeak: res.bufferPeak };
  }
  return { solvable: false, moves: [], bufferPeak: 0 };
}
